src/main/java/ch/vorburger/exec/AtomicExecuteResultHandler.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * #%L
+ * ch.vorburger.exec
+ * %%
+ * Copyright (C) 2012 - 2023 Michael Vorburger
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package ch.vorburger.exec;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.commons.exec.ExecuteException;
+import org.apache.commons.exec.ExecuteResultHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An {@link ExecuteResultHandler} which uses a single
+ * {@link java.util.concurrent.atomic.AtomicReference}
+ * instead of three separate volatile fields like the
+ * original {@link org.apache.commons.exec.DefaultExecuteResultHandler} did.
+ *
+ * @see <a href="https://github.com/vorburger/ch.vorburger.exec/issues/108">Issue #108</a>
+ */
+public class AtomicExecuteResultHandler implements ExecuteResultHandler {
```

Comment:
perhaps this is a naive question–this seems a bit like a CompletableFuture.  could we simplify this by using a CompletableFuture directly?

Reply:
I vaguely see what you mean, but not entirely. That would (just) be "nicer design" but not per-se "more (concurrency) safe", is that what you mean? I am open to a PR Contrib for it, but personally not motivated enough to spend more time to change this.

Reply:
Yeah, I think the main improvement would be a more precise `waitFor` method.  I'm not sure how much that matters, but it could speed things up slightly.  I'll take a stab at it if I get some free time